spec/datadog/core/configuration/option_spec.rb

```diff
@@ -221,6 +223,98 @@
+    # | rc, default                | rc                  | default         |
+    # | rc, default                | default             | rc              |
+    all_precedences = Datadog::Core::Configuration::Option::Precedence::LIST
+    (all_precedences.size + 1).times.flat_map { |s| all_precedences.combination(s).to_a }.each do |state|
+      context "for #{state.map { |s| s[1].to_s }} set" do
+        before do
+          # See this Option with many values set a different precedences.
+          state.each do |precedence|
+            # For convenience, the option value is set to the same object as the precedence.
+            value = precedence
+
+            @highest_value ||= value
+            option.set(value, precedence: precedence)
+          end
+        end
+
+        # Far all scenarios, try to remove each precedence and assert the correct behavior.
+        all_precedences.each do |precedence|
+          context "unsetting '#{precedence[1]}'" do
+            subject!(:unset) { option.unset(precedence) }
+            let(:precedence) { precedence }
+            let(:get) { option.get }
+
+            if state.empty?
+              context 'when no value is set' do
+                it 'resets the option' do
+                  expect(get).to eq(default)
+                  expect(option.send(:precedence_set)).to eq(Datadog::Core::Configuration::Option::Precedence::DEFAULT)
+                end
+              end
+            elsif precedence < state[0]
+              context 'when a value with lower precedence is unset' do
+                it 'does not modify the option value' do
+                  expect(get).to eq(@highest_value)
+                  expect(option.send(:precedence_set)).to eq(state[0])
+                end
+              end
+            elsif precedence == state[0]
+              context 'the highest precedence value is unset' do
+                if state.size == 1
+                  context 'removing the only value set' do
+                    it 'resets the option' do
+                      expect(get).to eq(default)
+                      expect(option.send(:precedence_set)).to eq(Datadog::Core::Configuration::Option::Precedence::DEFAULT)
+                    end
+                  end
+                else
+                  it 'falls back to lower precedence value' do
+                    expect(get).to eq(state[1])
+                    expect(option.send(:precedence_set)).to eq(state[1])
+                  end
+                end
+              end
+            elsif precedence > state[0]
+              context 'when a nonexistent value with higher precedence is unset' do
+                it 'does not modify the option value' do
+                  expect(get).to eq(@highest_value)
+                  expect(option.send(:precedence_set)).to eq(state[0])
+                end
+              end
+            end
+          end
+        end
+      end
+    end
+  end
```

Comment:
```suggestion
{
  empty_state: [],
  remote_configuration: [Datadog::Core::Configuration::Option::Precedence::REMOTE_CONFIGURATION],
  programatic: [Datadog::Core::Configuration::Option::Precedence::PROGRAMMATIC],
  default: [Datadog::Core::Configuration::Option::Precedence::DEFAULT],
  remote_and_programattic: [
    Datadog::Core::Configuration::Option::Precedence::REMOTE_CONFIGURATION,
    Datadog::Core::Configuration::Option::Precedence::PROGRAMMATIC
  ],
  remote_and_default: [
    Datadog::Core::Configuration::Option::Precedence::REMOTE_CONFIGURATION,
    Datadog::Core::Configuration::Option::Precedence::DEFAULT
  ],
  programatice_and_default: [
    Datadog::Core::Configuration::Option::Precedence::PROGRAMMATIC,
    Datadog::Core::Configuration::Option::Precedence::DEFAULT
  ],
  all: [
    Datadog::Core::Configuration::Option::Precedence::REMOTE_CONFIGURATION,
    Datadog::Core::Configuration::Option::Precedence::PROGRAMMATIC,
    Datadog::Core::Configuration::Option::Precedence::DEFAULT
  ]
}.each do |name, precedences|
    context "for #{name} set" do
      before do
        # See this Option with many values set a different precedences.
        precedences.each do |precedence|
          # For convenience, the option value is set to the same object as the precedence.
          value = precedence

          @highest_value ||= value
          option.set(value, precedence: precedence)
        end
      end

      # Far all scenarios, try to remove each precedence and assert the correct behavior.
      Datadog::Core::Configuration::Option::Precedence::LIST.each do |precedence|
        context "unsetting '#{precedence[1]}'" do
          subject!(:unset) { option.unset(precedence) }
          let(:precedence) { precedence }
          let(:get) { option.get }

          if precedences.empty?
            context 'when no value is set' do
              it 'resets the option' do
                expect(get).to eq(default)
                expect(option.send(:precedence_set)).to eq(Datadog::Core::Configuration::Option::Precedence::DEFAULT)
              end
            end
          elsif precedence < precedences[0]
            context 'when a value with lower precedence is unset' do
              it 'does not modify the option value' do
                expect(get).to eq(@highest_value)
                expect(option.send(:precedence_set)).to eq(precedences[0])
              end
            end
          elsif precedence == precedences[0]
            context 'the highest precedence value is unset' do
              if precedences.size == 1
                context 'removing the only value set' do
                  it 'resets the option' do
                    expect(get).to eq(default)
                    expect(option.send(:precedence_set)).to eq(Datadog::Core::Configuration::Option::Precedence::DEFAULT)
                  end
                end
              else
                it 'falls back to lower precedence value' do
                  expect(get).to eq(precedences[1])
                  expect(option.send(:precedence_set)).to eq(precedences[1])
                end
              end
            end
          elsif precedence > precedences[0]
            context 'when a nonexistent value with higher precedence is unset' do
              it 'does not modify the option value' do
                expect(get).to eq(@highest_value)
                expect(option.send(:precedence_set)).to eq(precedences[0])
              end
            end
          end
        end
      end
    end
  end
end
```

Cosmetic change comment 😄 

I find it more readable to see the list of combination that we are running on the tests upfront than for it to be dynamically generated. 

If we add a new precedence value, I know it has the issue that we manually have to come to add the combination. Not sure how often we do that, but for starters, I would go for readability than automation 😄 

As being a cosmetic suggestion is not a blocker for me

Reply:
I tried this at first, but thought it would look silly if we added one more precedence (say ENV_VAR), because it grows non-linearly.

But you know what, we don't need to test every combination, as a lot of them are redundant.
For 3 possible values, the redundancy is not too bad, but for 4 or more it there's a lot of waste.

Let me take in these changes.